userspace/engine/rule_loader.cpp

```diff
@@ -0,0 +1,1098 @@
+		"startswith", "endswith", "in", "intersects", "pmatch"};
+	return find(ops.begin(), ops.end(), op) != ops.end();
+}
+
+// todo: this should be in libsinsp
+static bool is_operator_for_list(std::string op)
+{
+	return op == "in" || op == "intersects" || op == "pmatch";
+}
+
+static bool is_format_valid(falco_engine* e, string src, string fmt, string& err)
+{
+	try
+	{
+		shared_ptr<gen_event_formatter> formatter;
+		formatter = e->create_formatter(src, fmt);
+		return true;
+	}
+	catch(exception &e)
+	{
+		err = e.what();
+		return false;
+	}
+}
+
+static string yaml_format_object(
+		const string& content,
+		const vector<YAML::Node>& docs,
+		vector<YAML::Node>::iterator doc,
+		YAML::iterator node)
+{
+	YAML::Node item = *node++;
+	YAML::Node cur_doc = *doc++;
+	// include the "- " sequence mark
+	size_t from = item.Mark().pos - 2;
+	size_t to = 0;
+	if (node != cur_doc.end())
+	{
+		// end of item is beginning of next item
+		to = node->Mark().pos - 2;
+	}
+	else if (doc != docs.end())
+	{
+		// end of item is beginning of next doc
+		to = doc->Mark().pos - 4;
+	}
+	else
+	{
+		// end of item is end of file contents
+		to = content.length();
+	}
+	string obj = content.substr(from, to - from);
+	return trim(obj);
+}
+
+template <typename T>
+static bool yaml_is_type(const YAML::Node& v)
+{
+	T t;
+	return v.IsDefined() && v.IsScalar() && YAML::convert<T>::decode(v, t);
+}
+
+template <typename T>
+static void yaml_decode_seq(const YAML::Node& item, vector<T>& out)
+{
+	T value;
+	for(const YAML::Node& v : item)
+	{
+		THROW(!v.IsScalar() || !YAML::convert<T>::decode(v, value),
+			"Can't decode YAML sequence value");
+		out.push_back(value);
+	}
+}
+
+template <typename T>
+static void yaml_decode_seq(const YAML::Node& item, set<T>& out)
+{
+	T value;
+	for(const YAML::Node& v : item)
+	{
+		THROW(!v.IsScalar() || !YAML::convert<T>::decode(v, value),
+			"Can't decode YAML sequence value");
+		out.insert(value);
+	}
+}
+
+// todo(jasondellaluce): this breaks string escaping in lists
+static bool resolve_list(string& cnd, YAML::Node& list)
+{
+	static string blanks = " \t\n\r";
+	static string delims = blanks + "(),=";
+	string new_cnd;
+	size_t start, end;
+	bool used = false;
+	start = cnd.find(list["list"].as<string>());
+	while (start != string::npos)
+	{
+		// the characters surrounding the name must
+		// be delims of beginning/end of string
+		end = start + list["list"].as<string>().length();
+		if ((start == 0 || delims.find(cnd[start - 1]) != string::npos)
+			&& (end >= cnd.length() || delims.find(cnd[end]) != string::npos))
+		{
+			// shift pointers to consume all whitespaces
+			while (start > 0
+				&& blanks.find(cnd[start - 1]) != string::npos)
+			{
+				start--;
+			}
+			while (end < cnd.length()
+				&& blanks.find(cnd[end]) != string::npos)
+			{
+				end++;
+			}
+			// create substitution string by concatenating all values
+			string sub = "";
+			for (auto v : list["items"])
```

Comment:
Another case where you could separate the parsing of the list items into an object, and just inject them into the string (or even later, ast) later.

Reply:
We agreed to defer a more concrete split of yaml parsing and other rule loading to a follow-on PR.